rfd/0109-cloud-agent-upgrades.md

```diff
@@ -327,6 +327,33 @@ to be changed to follow the same pattern of registering a release channel and in
 by ec2 discovery.
 
 
+### ECS Fargate Agents
+When Teleport Discover flow is used to connect an RDS, a Teleport agent is automatically provisioned using ECS Fargate. The automatic update
+functionality for regular agents depends on systemd or Kubernetes. Fargate agents run in containers, so that means the existing automatic
+update functionality is not compatible with these agents. We need a separate method of keeping these Fargate agents updated.
+
+To keep Farage agents updated, Teleport will run a separate task that checks every 30 minutes for an available update and uses the ECS
+Fargate API to update Teleport agents. This task does not need to be run on all Teleport clusters. It will only be needed if the
+`TELEPORT_AUTOMATIC_UPGRADES` environment variable is enabled. The window when Fargate agents are updated should also coincide with the
+tenant upgrade window.
+
+The logic flow will look like the same as the core update model:
```

Comment:
The issue is that if we deploy a bad version, we won't get any heartbeats.

I think we should rely entirely on ECS APIs, except for getting the region which we need to call any AWS API.
How can we get the region? Should we iterate over Teleport (RDS) DBs and create a set of AWS regions, and then call DescribeCluster for each region?

Reply:
> There's still no section that explains how we will determine that there are agents running in Fargate.

Marco mentioned that `ecs:DescribeClusters` and `ecs:DescribeServices` provides tags that may be used to help identify if the cluster or service was created by Teleport using the DeploymentService. Just added the additional info in the rfd.

> Also, I'm wondering about failure scenarios.

I was thinking we might be able to use `ecs:WaitUntilTasksRunning` and `ecs:WaitUntilTasksStopped` to handle new agent failure and graceful shutdown. I haven't tested if it would work yet, but I was thinking we'd be able to use these APIs to wait for a new agent to be healthy before deleting the existing agent.

> In addition to just checking agent heartbeats, should we also look at Fargate tasks and reconcile them against the cluster state?

I don't fully understand this question. Is this regarding how to check if an agent is in a healthy state?

Reply:
We were discussing the problem of getting all the scoped agents for a given cluster
Let's use the APIs you describe above
For getting the region, let's iterate over all Databases, filter for RDS ones and extract the `AWS.Region`.
For each unique Region, we will call the APIs (DescribeCluster/DescribeServices)

cc @r0mant